libcob/common.c

```diff
@@ -7804,17 +7829,54 @@ cob_expand_env_string (char *strval)
 				k++;
 			}
 			k--;
-		} else if (strval[k] == '$'
-		        && strval[k+1] == '$') {	/* Replace $$ with process-id */
-			j += sprintf (&env[j], "%d", cob_sys_getpid());
+		} else if (strval[k] == '$') {
+			struct cob_time time;
+			const char *s = NULL;
+		        switch ( strval[k+1] ){
+			case '$':	/* Replace $$ with process-id */
+				j += sprintf (&env[j], "%d", cob_sys_getpid());
+				k++;
+				continue;
+			case 'm': /* $m is the executable filename */
+				s = COB_MODULE_PTR->module_name;
+				break;
+			case 'f': /* $f is the executable filename */
+				s = cobglobptr->cob_main_argv0;
+				break;
+			case 'b': /* $b is the executable basename */
+				s = get_basename (cobglobptr->cob_main_argv0);
+				break;
+			case 'd': /* $d date as yyyymmdd */
+				time = cob_get_current_datetime (DTR_DATE);
+				j += sprintf (&env[j], "%04d%02d%02d",
+					      time.year, time.month,
+					      time.day_of_month);
+				k++;
+				continue;
+			case 't': /* $t time as hhmmss */
+				time = cob_get_current_datetime (DTR_TIME_NO_NANO);
+				j += sprintf (&env[j], "%02d%02d%02d",
+					      time.hour, time.minute, time.second);
+				k++;
+				continue;
+			default:
+				env[j++] = strval[k];
+				continue;
+			}
+			if (!s) s = "NULL";
+			copylen = strlen(s);
+			if (copylen + j > envlen - 128) {
+				env = cob_realloc (env, envlen, j + copylen + 256);
+				envlen = j + copylen + 256;
+			}
+			j += sprintf (&env[j], "%s", s);
 			k++;
```

Comment:
please leave the $$ code and the $something "intact" and copy/adjust that for all $d and $t, which allows to drop the current use of `continue`

Reply:
I changed the code, to remove the `continue`, but I kept the `switch`, that I think makes the code more readable.